src/main/java/com/faforever/client/login/OAuthValuesReceiver.java

```diff
@@ -0,0 +1,134 @@
+package com.faforever.client.login;
+
+import com.faforever.client.config.ClientProperties;
+import com.faforever.client.fx.JavaFxUtil;
+import com.faforever.client.preferences.PreferencesService;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+@RequiredArgsConstructor
+@Slf4j
+public class OAuthValuesReceiver {
+
+  private static final Pattern CODE_PATTERN = Pattern.compile("code=([^ &]+)");
+  private static final Pattern STATE_PATTERN = Pattern.compile("state=([^ &]+)");
+  private static final List<String> ALLOWED_HOSTS = List.of("localhost");
+
+  private final CompletableFuture<URI> addressFuture = new CompletableFuture<>();
+
+  private final ClientProperties clientProperties;
+  private final PreferencesService preferencesService;
+
+  public CompletionStage<Values> receiveValues() {
+    return CompletableFuture.supplyAsync(() -> {
+      Collection<URI> redirectUris = retrieveRedirectUris();
+
+      for (URI uri : redirectUris) {
+        try {
+          return readWithUri(uri);
+        } catch (SocketException e) {
+          log.warn("Port " + uri.getPort() + " is probably already in use", e);
+        }
+      }
+      throw new IllegalStateException("Could not read from any redirect URI: " + redirectUris);
+    });
+  }
+
+  @SneakyThrows
+  private List<URI> retrieveRedirectUris() {
+    JavaFxUtil.assertBackgroundThread();
+    if (clientProperties.isUseRemotePreferences()) {
+      return preferencesService.getRemotePreferences().getOAuth().getRedirectUris().stream()
+          .filter(uri -> uri.getPort() > 0)
+          .filter(uri -> ALLOWED_HOSTS.contains(uri.getHost()))
+          .collect(Collectors.toList());
+    }
+    return IntStream.range(60000, 65535)
+        .mapToObj(port -> URI.create("http://localhost:%d/".formatted(port)))
+        .collect(Collectors.toList());
```

Comment:
Does this just scan all possible ports? What use case is this for?

Reply:
If remote preferences aren't used, the server can't tell us which redirect URIs/ports are whitelisted (as in the comment above). This is the case when running the client in offline mode (where there is no OAuth anyway) or in unit-tests where the port doesn't matter; the list just needs to be long enough so that a free port can be found.

Spring provides `org.springframework.util.SocketUtils#findAvailableTcpPort()` which would seem perfect, however, from experience I know that it is not reliable: it tries to open a port and if that's successful, it will close the port and return its number, assuming that it will still be free when the caller uses it to open a port - for which there is no guarantee.

I added a comment.

Reply:
If this is based around unit tests then wold it not be better to define the port in the test and then mock the client configuration to return the uris with the opened port. And in the case of us being offline then this should never be called anyway.

Reply:
How can a unit test know which port will be free to open?

Reply:
We can just create the same list of ports in the unit test and hand it over by mocking PreferencesService

Reply:
Fair enough. And what should the `OAuthValuesReceiver` do if `clientProperties.isUseRemotePreferences()` is false? That doesn't necessarily translate to offline mode.

Reply:
In that case the values would have to be input into a nominal redirect_uris textbox like the other endpoints are on the login screen. So a method to specify those manually needs to be added.

Reply:
I added it to the login form